train.py

```diff
@@ -0,0 +1,193 @@
+            loss = nn.MSELoss()
+        elif self.hparams["loss_type"] == "mae":
+            loss = nn.L1Loss()
+        elif self.hparams["loss_type"] == "huber":
+            loss = nn.HuberLoss()
+        else:
+            raise NotImplementedError()
+
+        return loss
+
+    def _save_models(self):
+        # ckpt specifies directory and name of the file is name of the experiment in wandb
+        save_path = f"{self.hparams['ckpt_dir']}/{self.logger.experiment.name}.ckpt"
+        # saving models
+        torch.save({"model": self.model.state_dict(), "forward_diffusion": self.forward_diffusion.state_dict()}, save_path)
+
+    def forward(self, x: torch.Tensor, t: torch.Tensor) -> torch.Tensor:
+        return self.model(x, t)
+
+    def configure_optimizers(self) -> optim.Optimizer:
+        optimizer = optim.AdamW(self.model.parameters(), lr=self.hparams["lr"], weight_decay=self.hparams["l2"])
+
+        return optimizer
+
+    def training_step(self, batch: torch.Tensor, batch_idx: int) -> float:
+        x = batch["x"]
+
+        batch_size = x.shape[0]
+
+        # sample t
+        t = torch.randint(0, self.timesteps, size=(batch_size,), dtype=torch.long, device=self.device)
+
+        # noise batch
+        x_noisy, added_noise = self.forward_diffusion.forward(x, t)
+
+        # get predicted noise
+        predicted_noise = self.model.forward(x_noisy, t)
+
+        # get loss value for batch
+        loss = self.loss_function_module(predicted_noise, added_noise)
+
+        self.log("loss", loss, on_epoch=True)
+
+        if self.global_step % self.hparams["save_every_n_steps"] == 0:
+            self._save_models()
+
+        return loss
+
+
+def preprocess_dataset(filepath: str, batch_size: int, num_workers: int):
+    dataset = load_dataset(filepath)
+    transformed_dataset = dataset.with_transform(transforms).remove_columns("label")
+
+    dataloader = DataLoader(transformed_dataset["train"], batch_size=batch_size, num_workers=num_workers)
+
+    return dataloader
+
+
+def transforms(examples):
+    transform = Transforms(img_size=28)
+
+    examples["x"] = [transform.img2torch(image.convert("L")) for image in examples["image"]]
+    del examples["image"]
+
+    return examples
+
+
+def makedir_if_not_exists(dir: str):
+    if not os.path.exists(dir):
+        os.makedirs(dir)
+
+
+@hydra.main(config_path="configs", config_name="config-fashion-mnist")
+def train(cfg: Config):
+    # get dataloader
+    dataloader = preprocess_dataset(
+        cfg.hyperparameters.dataset, batch_size=cfg.hyperparameters.batch_size, num_workers=cfg.hyperparameters.num_workers
+    )
+
+    # generate some random id for the run
+    run_id = str(uuid.uuid1())[:8]
+
+    # make directory for logs if doesn't exist
+    makedir_if_not_exists(cfg.paths.log_dir)
+
+    # initializing logger
+    logger = WandbLogger(project="annotated-diffusion", name=f"{cfg.logger.run_name}-{run_id}", save_dir=cfg.paths.log_dir)
```

Comment:
```suggestion
    # make directory for logs if doesn't exist
    makedir_if_not_exists(cfg.paths.log_dir)

    # initializing logger
    logger = WandbLogger(
      project="annotated-diffusion",
      name=cfg.name,
      save_dir=cfg.paths.log_dir,
    )
```

I recommend using a timestamp-based id rather than random one - it makes tracking runs much easier. You can also get this from hydra, by setting something like this in your config:

```yaml
name: ${now:%Y_%m_%d_%H_%M}_diffusion
```